### ISSUE TITLE
syntax/java.vim markdown hl group name

### ISSUE THREAD

**Issue by Shane-XB-Qian** [vim/vim#16349]
### Steps to reproduce

Error detected while processing BufRead Autocommands for "*.java"..FileType Autocommands for "*"..Syntax Autocommands for "*"..function <SNR>34_SynSet[25]..script /usr/local/share/vim/vim91/syntax/java.vim:
line  425:
E28: No such highlight group name: markdownLineStart markdownH1 markdownH2 markdownHeadingRule markdownRule markdownCode markdownCodeBlock markdownIdDeclaration

### Expected behaviour

no err

### Version of Vim

v9.1.980

### Environment

linux

**Comment by chrisbra** [comment]
Is there something special with your installation? The java syntax script should have included the markdown runtime file, which then should have defined those groups: ` markdownLineStart markdownH1 markdownH2 markdownHeadingRule markdownRule markdownCode markdownCodeBlock markdownIdDeclaration`? Are you using a different `syntax/markdown.vim`  plugin possibly?

**Comment by Shane-XB-Qian** [comment]
yea, possible, some markdown.vim, but:
`    syn clear markdownId markdownLineStart markdownH1 markdownH2 markdownHeadingRule markdownRule markdownCode markdownCodeBlock markdownIdDeclaration
` 
at line 425 of java.vim, not `syn clear` perhaps help on this.

**Comment by zzzyxwvut** [comment]
Do `java.vim` and `markdown.vim` come from `$VIMRUNTIME`?

What does `:filter /syntax/ scriptnames` print _after_ loading
any Java source file?

**Comment by Shane-XB-Qian** [comment]
it happened when _enter_ java file, so the err is annoyed.
if ok, pls not 'syn clear'

-- 
shane.xb.qian

**Comment by zzzyxwvut** [comment]
If you prefer using a _non-distributed_ `markdown.vim`, simply  
disable the recognition of Markdown comments in Java files  
with this setting in `vimrc`:

```vim
let g:java_ignore_markdown = 1
```

**Comment by Shane-XB-Qian** [comment]
> If you prefer using a _non-distributed_ `markdown.vim`, simply  
> disable the recognition of Markdown comments in Java files  
> with this setting in `vimrc`:
> 
> ```vim
> let g:java_ignore_markdown = 1
> ```

i knew this workaround, but perhaps java.vim should not 'syn clear' markdown.vim 's hl group?
// and there were just some fs same named 'markdown.vim', that's possible, not only 'markdown.vim'
// but not means i donot use _non-distributed_ markdown.vim

-- 
shane.xb.qian

**Comment by zzzyxwvut** [comment]
The support for Markdown comments is written against the  
distributed `syntax/markdown.vim`.  The cleared highlighting  
groups are *redefined* to match significant leading whitespace  
characters *in* `///` comments, whereas their `:syn-include`d  
definitions know nothing about `///` comments and always match  
significant leading whitespace characters at `^`.  Chances are  
that the other `markdown.vim` scripts are no different in this  
regard.

**Comment by Shane-XB-Qian** [comment]
how about applying your change to @tpope ? is it better?

**Comment by chrisbra** [comment]
So perhaps we should use this patch instead @zzzyxwvut ?
```
diff --git a/runtime/syntax/java.vim b/runtime/syntax/java.vim
index b3e17b55f..60bb9c124 100644
--- a/runtime/syntax/java.vim
+++ b/runtime/syntax/java.vim
@@ -381,7 +381,7 @@ if !exists("g:java_ignore_javadoc") && (s:with_html || s:with_markdown) && g:mai
   " Include HTML syntax coloring for Javadoc comments.
   if s:with_html
     try
-      syntax include @javaHtml syntax/html.vim
+      syntax include @javaHtml $VIMRUNTIME/syntax/html.vim
     finally
       unlet! b:current_syntax
     endtry
@@ -390,7 +390,7 @@ if !exists("g:java_ignore_javadoc") && (s:with_html || s:with_markdown) && g:mai
   " Include Markdown syntax coloring (v7.2.437) for Javadoc comments.
   if s:with_markdown
     try
-      syntax include @javaMarkdown syntax/markdown.vim
+      syntax include @javaMarkdown $VIMRUNTIME/syntax/markdown.vim
       let s:ff.WithMarkdown = s:ff.LeftConstant
     catch /\<E48[45]:/
       call s:ReportOnce(v:exception)
```

**Comment by zzzyxwvut** [comment]
Java users of [`vim-polyglot`](https://github.com/sheerun/vim-polyglot) and similar plugins that offer  
alternative runtime files for Markdown and/or HTML but not  
for Java should _opt out_ of Javadoc comments as documented:

https://github.com/vim/vim/blob/8dc98bf427d62501d9ecbb48d20ae2633c3db187/runtime/doc/syntax.txt#L2126-L2131

as long as alternative files do not extend distributed files  
and their syntax items bear distinct names (compare [`mkdCode`](https://github.com/sheerun/vim-polyglot/blob/f5393cfee07aeb666f4d75f9b3a83163862fb094/syntax/markdown.vim#L105)  
and [`markdownCode`](https://github.com/vim/vim/blob/8dc98bf427d62501d9ecbb48d20ae2633c3db187/runtime/syntax/markdown.vim#L121)).

We shouldn't inconvenience neither these users nor users of  
distributed runtime files with confining related syntax file  
lookups to `$VIMRUNTIME`.  People will come here and complain  
that their `&rtp` setups are no longer respected _without their  
fiddling with_ `$VIMRUNTIME` or worse -- that they now have to  
move their symlinks that point to latest versions in their  
cloned upstream repositories out of `~/.vim/syntax` etc. _to_  
`$VIMRUNTIME`.

**Comment by chrisbra** [comment]
alright, let's close it then as invalid.

**Comment by Shane-XB-Qian** [comment]
um? so the argument is acceptable? which Java.vim syn clear markdown.vim 's hl group? :shrug:

--
shane.xb.qian

**Comment by chrisbra** [comment]
Yes for a specific reason. And you didn't even tell us about markdown.vim you are using. Perhaps @zzzyxwvut wants to add some try/catch statement around the syn clear command

**Comment by zzzyxwvut** [comment]
@Shane-XB-Qian, there are no failed [CI jobs](https://github.com/vim/vim/actions/runs/12544657610) for `v9.1.0980`.
Launching Vim with `-clean`:

```shell
vim -clean Foo.java
```

should work.

If you prefer using some third-party `syntax/markdown.vim`  
and see errors, opt out:

```shell
vim --cmd 'let g:java_ignore_markdown = 1' Foo.java
```

**Comment by Shane-XB-Qian** [comment]
I think if you're going to search there lot markdown.vim, but should be only one in the vim distributed repo, and that one should own the hl group. simple and clear~ and actually that's a question not only to markdown.vim, just well knew Java.vim is doing this way.

--
shane.xb.qian

**Comment by Shane-XB-Qian** [comment]
yea, try/catch maybe anthor way to avoid this, if really to do such syn clear.

--
shane.xb.qian

**Comment by zzzyxwvut** [comment]
Alright, let's catch and log `E28` (in case the support for  
Markdown comments is desired but the sourcing order prevents  
it) and _enforce opting out_.

**Comment by zzzyxwvut** [comment]
Try the Java syntax file version from #16369.  You shall see  
a caught and logged `E28` exception for any *first* buffer  
loaded with a Java source file and `g:java_ignore_markdown`  
shall be set to `28` from now on.

Should you still find it annoying, you know what to do.

**Comment by Shane-XB-Qian** [comment]
ah~ i felt you made it complex, i think user there had alternative syntax files (for some ft) is possible, not markdown.vim, or html.vim, if you not syn clear it (at your java.vim) it simply vim chose one to invoke or no err happened;
but ok, now i felt the err maybe worthy too, (vs) at least no worthy to make it more complex, so let it be~

**Comment by zzzyxwvut** [comment]
What other filetypes but Markdown are on your mind?

**Only for the current Java buffer**, a subset of Markdown  
syntax groups is cleared with `:syn clear markdownId ...`.   
For the reference:

https://github.com/vim/vim/blob/663d18d6102f40d14e36096ec590445e61026ed6/runtime/doc/syntax.txt#L6018-L6020

If there are NO Markdown comments `///` in a Java buffer, it  
doesn't matter where the runtime files are coming from when  
the patch is applied _or_ `g:java_ignore_markdown` is defined.   
You can load an arbitrary `*.markdown` or `*.md` file in a new  
buffer and depend on a third-party `markdown.vim` for syntax  
highlighting, without observing any effect from the syntax  
clearing done in any _Java_ buffer.

If there are Markdown comments `///` in a Java buffer and you  
want such comments recognised, you have to make sure that no  
third-party `markdown.vim` is loaded before the distributed  
`markdown.vim` (with `v9.1.0750` and newer).  As you already  
know, there is no _Markdown comments_ support for third-party  
runtime files.
